### PR TITLE
Support named PK constraints

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Resolves https://github.com/crate/crate/issues/11025 which is to allow PK constraints to be named. PK names can be handled the same way as the names for `check` constraints. The only difference should be that the PK is unique per table. When PK names are provided, they will be persisted in `meta` mapping with `primary_key_constraint_name` as the key.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
retest this please
unrelated test failed

**Comment by jeeminso** [comment]
Noticed https://github.com/crate/crate/blob/bf3eab10bbab3096a93fcb17cfa4cefd04374e4f/server/src/main/java/io/crate/execution/ddl/tables/TableCreator.java#L105, will do this after merging this PR.

**Comment by mfussenegger** [comment]
> Noticed
> 
> https://github.com/crate/crate/blob/bf3eab10bbab3096a93fcb17cfa4cefd04374e4f/server/src/main/java/io/crate/execution/ddl/tables/TableCreator.java#L105
> , will do this after merging this PR.

If there is no cost in keeping it we don't need to remove it.  
It would break create table in mixed version clusters with old nodes. People sometimes skip versions,  so supporting this a bit longer is an upside

**Review by matriv — COMMENTED** [review]
Thank you! Left some minor comments.
Additionally:
1. we need to add the pk to `pg_constraint` table: https://www.postgresql.org/docs/current/catalog-pg-constraint.html
2. Maybe update more docs:
    - `general/constraints.rst`: maybe provide the syntax the same way we do for check and provide examples
    - `update ddl/create-table.rst`: check and maybe update docs, provide examples (not only the syntax reference)

**Comment by mfussenegger** [comment]
> we need to add the pk to pg_constraint table: https://www.postgresql.org/docs/current/catalog-pg-constraint.html

This can be done in a follow up PR. We've basic exposure via `information_schema.table_constraints` 
Please don't increase the scope of PRs if not necessary

**Comment by matriv** [comment]
> > we need to add the pk to pg_constraint table: https://www.postgresql.org/docs/current/catalog-pg-constraint.html
> 
> This can be done in a follow up PR. We've basic exposure via `information_schema.table_constraints` Please don't increase the scope of PRs if not necessary

Of course I don't mind if this is done as a followup, but I thought I should mention as it's missing, and to me it seems to "belong" to the same scope.

**Comment by jeeminso** [comment]
Thanks for reviewing @matriv, Please check again. 
Regarding `ddl/create-table.rst`, I did not find much information regarding `primary keys` nor `constraints` in general so I did not update it.

File: server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java [hunk 1]
```diff
@@ -124,6 +128,20 @@ public BoundCreateTable bind(NumberOfShards numberOfShards,
...
+        String pkConstraintName = null;
+        if (!pkConstraintNames.isEmpty()) {
+            // CrateDB allows 'create table t (a int constraint c1 primary key, b int constraint c2 primary key);',
+            // making sure that c1 == c2.
+            if (pkConstraintNames.size() > 1) {
+                throw new IllegalArgumentException(
+                    "More than one name for PRIMARY KEY constraint provided: " + String.join(",", pkConstraintNames));
+            }
+            pkConstraintName = pkConstraintNames.iterator().next();
+            if ("".equals(pkConstraintName)) {
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java:140]: `pkConstraintName` can be a `null` if not provided. It just cannot be an empty string.

File: docs/sql/general/constraints.rst [hunk 2]
```diff
@@ -70,21 +111,21 @@ satisfy a :ref:`boolean expression <sql-literal-value>` on ``INSERT`` and
 
 Syntax::
 
-  [CONSTRAINT check_name>] CHECK (boolean_expression)
+  [CONSTRAINT <check_name>] CHECK (boolean_expression)
 
-If ``CONSTAINT check_name`` is omitted, CrateDB generates a unique name
+If ``CONSTAINT <check_name>`` is omitted, CrateDB generates a unique name
```
Comment (matriv) [inline docs/sql/general/constraints.rst:116]:
```suggestion
If ``CONSTRAINT <check_name>`` is omitted, CrateDB generates a unique name
```

File: libs/sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java [hunk 3]
```diff
@@ -748,6 +750,18 @@ public void testCreateTableStmtBuilder() {
         printStatement("create table test (col1 string storage with (columnstore = false))");
     }
 
+    @Test
+    public void test_named_primary_key_constraint_without_name_throws() {
```
Comment (matriv) [inline libs/sql-parser/src/test/java/io/crate/sql/parser/TestStatementBuilder.java:754]:
```suggestion
    public void test_named_primary_key_constraint_without_name_is_not_allowed() {
```

File: server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java:140]: of course, sorry, mixed it up with the Set.

File: server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java [hunk 4]
```diff
@@ -124,6 +128,20 @@ public BoundCreateTable bind(NumberOfShards numberOfShards,
...
+                throw new IllegalArgumentException(
+                    "The name of primary key constraint must not be empty, please try without CONSTRAINT keyword");
```
Comment (matriv) [inline server/src/main/java/io/crate/analyze/AnalyzedCreateTable.java:142]: Just a suggestion:
```suggestion
                    "The name of primary key constraint must not be empty, please either use a name or remove the CONSTRAINT keyword");
```

File: docs/sql/general/constraints.rst [hunk 5]
```diff
@@ -33,6 +33,47 @@ can be declared as ``PRIMARY KEY`` instead.
 
 Adding a ``PRIMARY KEY`` column is only possible if the table is empty.
 
+Syntax::
+
+  [CONSTRAINT <name>] PRIMARY KEY [ column_name [, ... ] ]
+
+For example, a table with a named ``PRIMARY KEY`` constraint can be create like
```
Comment (matriv) [inline docs/sql/general/constraints.rst:40]: Maybe:
```suggestion
For example, a table with a named ``PRIMARY KEY`` constraint can be created with::
```
and remove `below::`

File: docs/sql/general/constraints.rst [hunk 6]
```diff
@@ -33,6 +33,47 @@ can be declared as ``PRIMARY KEY`` instead.
...
+
+The named ``PRIMARY KEY`` constraints can be inlined::
+
+    cr> CREATE TABLE person2 (
+    ...     firstname TEXT CONSTRAINT c PRIMARY KEY,
+    ...     lastname TEXT CONSTRAINT c PRIMARY KEY
+    ... );
+    CREATE OK, 1 row affected  (... sec)
+
+If another column is required to be added as ``PRIMARY KEY`` column::
```
Comment (matriv) [inline docs/sql/general/constraints.rst:58]:
```suggestion
If a new column is required to be added with ``PRIMARY KEY`` constraint::
```

File: docs/sql/general/constraints.rst [hunk 7]
```diff
@@ -33,6 +33,47 @@ can be declared as ``PRIMARY KEY`` instead.
...
+
+    cr> ALTER TABLE person2 ADD COLUMN middleName PRIMARY KEY;
+    ALTER OK, -1 rows affected  (... sec)
+
+The ``PRIMARY KEY`` constraint can be unnamed like below::
```
Comment (matriv) [inline docs/sql/general/constraints.rst:63]:
```suggestion
The ``PRIMARY KEY`` constraint can also be unnamed, e.g.::
```

**Review by matriv — APPROVED** [review]
LGTM. Thx a lot @jeeminso !
Left some comments about wording suggestions, which you don't need to adopt, but check them, also a couple of typos.

File: docs/sql/general/constraints.rst [hunk 8]
```diff
@@ -33,6 +33,47 @@ can be declared as ``PRIMARY KEY`` instead.
 
 Adding a ``PRIMARY KEY`` column is only possible if the table is empty.
 
+Syntax::
+
+  [CONSTRAINT <name>] PRIMARY KEY [ column_name [, ... ] ]
+
+For example, a table with a named ``PRIMARY KEY`` constraint can be create
```
Comment (matriv) [inline docs/sql/general/constraints.rst:40]:
```suggestion
For example, a table with a named ``PRIMARY KEY`` constraint can be created
```

Comment (jeeminso) [reply docs/sql/general/constraints.rst:40]: Thank you!

**Comment by jeeminso** [comment]
```
cr> create table t (a int constraint c_123 primary key);                                                                                    
CREATE OK, 1 row affected  (1.883 sec)
cr> select conname from pg_catalog.pg_constraint where conname = 'c_123';                                                                   
+---------+
| conname |
+---------+
| c_123   |
+---------+
SELECT 1 row in set (0.026 sec)
```

It looks like `pg_catalog.pg_constraint` is implicitly taken care of.

**Comment by matriv** [comment]
thx for checking!  then, you can just add a test that tests an explicit pk name.

**Comment by mfussenegger** [comment]
> Of course I don't mind if this is done as a followup, but I thought I should mention as it's missing, and to me it seems to belong to the same scope.

Oh sorry, I read that as needing to add the pg_constraint table. I forgot we already have it.

**Comment by matriv** [comment]
> Oh sorry, I read that as needing to add the pg_constraint table. I forgot we already have it.
No worries!

**Pull request merged**